packages/remix-server-runtime/server.ts

```diff
@@ -265,7 +302,214 @@ async function handleDataRequestRR(
+          new URL(request.url).searchParams.get("_routes"),
+          staticHandler,
+          matches,
+          loadContext,
+          handleError,
+          serverMode,
+          build
+        );
+
+  // Mark all successful responses with a header so we can identify in-flight
+  // network errors that are missing this header
+  let resultHeaders = new Headers(headers);
+  resultHeaders.set("X-Remix-Response", "yes");
+  resultHeaders.set("Content-Type", "text/x-turbo");
+
+  // Note: Deferred data is already just Promises, so we don't have to mess
+  // `activeDeferreds` or anything :)
+  return new Response(
+    encode(result, [
+      (value) => {
+        if (value instanceof ErrorResponseImpl) {
+          return ["ErrorResponse", { ...value }];
+        }
+      },
+    ]),
+    { headers: resultHeaders }
+  );
+}
+
+async function singleFetchAction(
+  request: Request,
+  handlerUrl: URL,
+  staticHandler: StaticHandler,
+  loadContext: AppLoadContext,
+  handleError: (err: unknown) => void
+): Promise<[SingleFetchResult, Headers]> {
+  try {
+    let handlerRequest = new Request(handlerUrl, {
+      method: request.method,
+      body: request.body,
+      headers: request.headers,
+      signal: request.signal,
+      ...(request.body ? { duplex: "half" } : undefined),
+    });
+    let response = await staticHandler.queryRoute(handlerRequest, {
+      requestContext: loadContext,
+    });
+    // callRouteLoader/callRouteAction always return responses
+    invariant(
+      isResponse(response),
+      "Expected a Response to be returned from queryRoute"
+    );
+    if (isRedirectResponse(response)) {
+      return [
+        {
+          redirect: response.headers.get("Location")!,
+          status: response.status,
+          revalidate: response.headers.has("X-Remix-Revalidate"),
+          reload: response.headers.has("X-Remix-Reload-Document"),
+        },
+        response.headers,
+      ];
+    }
+    return [
+      { data: await unwrapResponse(response), status: response.status },
+      response.headers,
+    ];
+  } catch (err) {
+    handleError(err);
+    let error = isResponse(err)
+      ? new ErrorResponseImpl(
+          err.status,
+          err.statusText,
+          await unwrapResponse(err)
+        )
+      : err;
+    return [{ error }, new Headers()];
+  }
+}
+
+async function singleFetchLoaders(
+  handlerUrl: URL,
+  routesToLoad: string | null,
+  staticHandler: StaticHandler,
+  matches: RouteMatch<ServerRoute>[] | null,
+  loadContext: AppLoadContext,
+  handleError: (err: unknown) => void,
+  serverMode: ServerMode,
+  build: ServerBuild
+): Promise<[SingleFetchResults, Headers]> {
+  let context: StaticHandlerContext;
+  try {
+    let handlerRequest = new Request(handlerUrl);
+    let loadRouteIds = routesToLoad ? routesToLoad.split(",") : undefined;
+
+    let result = await staticHandler.query(handlerRequest, {
+      requestContext: loadContext,
+      loadRouteIds,
+    });
```

Comment:
Single fetch loader calls use `query` with an optional `loadRouteIds` if fine-grained revalidation is used